### PR TITLE
Support for uploading scoreDefinition and ScoreCardDefinition

### PR DESCRIPTION
**Pull request by shrutiparabgoogle**
Summary of my changes:
Each of the following changes are in a separate commit.

* Refactor: Move common utilities from controller to new package "patterns" (This refactor will be useful for future implementation, it is not immediately useful in this PR)
This refactor is required because there are a lot of controller utilities which can be reused for scoring. Example: reading patterns like `$resource.spec/artifacts/lint-spectral`. Hence, those common functions are moved to a separate package `patterns`. Since we planning to eventually get rid of the `core` package, these utils were not put there intentionally.
* Support for `ScoreDefinition` and `ScoreCardDefinition` upload: 
   -  TODO: yaml uploads
   - TODO: validation functions for definitions 
* Support for `ScoreDefinition` and `ScoreCardDefinition` print/get
* Support for `Score` and `ScoreCard` upload: 
   -  TODO: yaml uploads
* Support for `Score` and `ScoreCard` print/get

### CODE REVIEW

File: cmd/registry/cmd/upload/artifact_test.go [hunk 1]
```diff
@@ -62,32 +90,182 @@ func TestManifestArtifactUpload(t *testing.T) {
...
+	}{
+		{
+			desc:     "simple scorecard definition artifact upload",
+			project:  "upload-scorecard-definition-artifact-demo",
+			filePath: filepath.Join("testdata", "scorecard-definition.yaml"),
+			want: &rpc.ScoreCardDefinition{
+				Id:          "test-scorecard-definition",
+				Kind:        "ScoreCardDefinition",
+				DisplayName: "",
+				Description: "",
```
Comment (seaneganx) [inline cmd/registry/cmd/upload/artifact_test.go:246]: I think these can be left out.

**Review by seaneganx — APPROVED** [review]

File: cmd/registry/cmd/upload/artifact_test.go (hunk 1 above)
Comment (shrutiparabgoogle) [reply cmd/registry/cmd/upload/artifact_test.go]: Thanks for catching

File: cmd/registry/controller/controller.go [hunk 2]
```diff
@@ -109,8 +108,13 @@ func generateDependencyMap(
...
 	sourceMap := make(map[string]time.Time)
 
+	resourceName, err := patterns.ParseResourcePattern(resourcePattern)
+	if err != nil {
+		return nil, err
+	}
+
 	// Extend the dependency pattern if it contains $resource.api like pattern
-	extDependencyQuery, err := extendDependencyPattern(resourcePattern, dependency.Pattern, projectID)
+	extDependencyQuery, err := patterns.SubstitueReferenceEntity(dependency.Pattern, resourceName, projectID)
```
Comment (timburks) [inline cmd/registry/controller/controller.go:117]: Nit: "Substitue" should be "Substitute"

File: cmd/registry/patterns/parser.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	// Then try to match resource names.
+	resource, err = parseResource(resourcePattern)
+	if err == nil {
+		return resource, nil
+	}
+
+	return nil, fmt.Errorf("invalid resourcePattern: %s", resourcePattern)
+}
+
+func SubstitueReferenceEntity(resourcePattern string, referred ResourceName, projectID string) (string, error) {
```
Comment (timburks) [inline cmd/registry/patterns/parser.go:64]: nit: misspelling (noted above)

File: cmd/registry/patterns/resources.go [hunk 4]
```diff
@@ -0,0 +1,362 @@
+// Copyright 2021 Google LLC
```
Comment (timburks) [inline cmd/registry/patterns/resources.go:1]: Nit: new files should be copyright 2022

**Review by timburks — APPROVED** [review]
Just a couple of minor formatting/spelling issues - from another conversation, I think these protos might not be final, but I'm ok with making that update in a followup.

File: cmd/registry/cmd/upload/artifact.go [hunk 5]
```diff
@@ -92,6 +92,14 @@ func buildArtifact(ctx context.Context, parent string, filename string) (*rpc.Ar
 		artifact, err = buildReferenceListArtifact(ctx, jsonBytes)
 	case "TaxonomyList", patch.TaxonomyListMimeType:
 		artifact, err = buildTaxonomyListArtifact(ctx, jsonBytes)
+	case "ScoreDefinition", patch.ScoreDefinitionMimeType:
```
[the remark this answers is not in the document]
Comment (shrutiparabgoogle) [reply cmd/registry/cmd/upload/artifact.go]: Done.

File: cmd/registry/controller/controller.go (hunk 2 above)
Comment (shrutiparabgoogle) [reply cmd/registry/controller/controller.go]: Done

File: cmd/registry/patterns/parser.go (hunk 3 above)
Comment (shrutiparabgoogle) [reply cmd/registry/patterns/parser.go]: Done

File: cmd/registry/patterns/resources.go (hunk 4 above)
Comment (shrutiparabgoogle) [reply cmd/registry/patterns/resources.go]: Done

**Comment by shrutiparabgoogle** [comment]
> Just a couple of minor formatting/spelling issues - from another conversation, I think these protos might not be final, but I'm ok with making that update in a followup.

Yes, the proto update will be in a follow-up PR

**Pull request merged**